localstack/http/proxy.py

```diff
@@ -0,0 +1,151 @@
+from io import BytesIO
+from typing import Mapping, Union
+
+from werkzeug import Request, Response
+from werkzeug.datastructures import Headers
+from werkzeug.test import EnvironBuilder
+
+from .client import HttpClient, SimpleRequestsClient
+from .request import restore_payload
+
+
+def forward(
+    request: Request,
+    forward_base_url: str,
+    forward_path: str = None,
+    headers: Union[Headers, Mapping[str, str]] = None,
+) -> Response:
+    """
+    Convenience method that creates a new Proxy and immediately calls proxy.forward(...). See ``Proxy`` for more
+    information.
+    """
+    return Proxy(forward_base_url=forward_base_url).forward(
+        request, forward_path=forward_path, headers=headers
+    )
+
+
+class Proxy:
+    def __init__(self, forward_base_url: str, client: HttpClient = None):
+        """
+        Creates a new HTTP Proxy which can be used to forward incoming requests according to the configuration.
+
+        :param forward_base_url: the base url (backend) to forward the requests to.
+        :param client: the HTTP Client used to make the requests
+        """
+        self.forward_base_url = forward_base_url
+        self.client = client or SimpleRequestsClient()
+
+    def forward(
+        self,
+        request: Request,
+        forward_path: str = None,
+        headers: Union[Headers, Mapping[str, str]] = None,
+    ):
+        """
+        Uses the client to forward the given request according to the proxy's configuration.
+
+        :param request: the base request to forward (with the original URL and path data)
+        :param forward_path: the path to forward the request to. if set, the original path will be replaced completely,
+            otherwise the original path will be used
+        :param headers:
+        :return:
+        """
+        headers = Headers(headers) if headers else Headers()
+
+        if client_ip := request.remote_addr:
+            if xff := request.headers.get("X-Forwarded-For"):
+                headers["X-Forwarded-For"] = f"{xff}, {client_ip}"
+            else:
+                headers["X-Forwarded-For"] = f"{client_ip}"
+
+        if forward_path is None:
+            forward_path = request.path
+        if forward_path:
+            forward_path = "/" + forward_path.lstrip("/")
+
+        proxy_request = _copy_request(request, self.forward_base_url, forward_path, headers)
+        return self.client.request(proxy_request)
+
+
+class ForwardingHandler:
+    """
+    A dispatcher Handler which can be used in a ``Router[Handler]``  that proxies incoming requests according to the
+    configuration.
+    """
+
+    def __init__(self, proxy: Proxy, forward_path_parameter: bool = True):
```

Comment:
nit: Imho, it's quite hard to understand what the `forward_path_parameter` is doing (especially if you don't know too much about Werkzeug's route parameter converters). Maybe we could add a bit more documentation here? Or refer to the docs in `path_forwarder` (where there's a nice example with an nginx pendant)?

Reply:
yes good point, will add!

Reply:
actually decided to remove it all together to simplify things :-)